### DISCUSSION TITLE
Deepness - roadmap

### DISCUSSION THREAD

**Discussion by bartoszptak** [PUTvision/qgis-plugin-deepness#151] [Ideas]
Hi! We introduce the project roadmap! :fire: 

Here, you can suggest new features and improvements!

:fire: :fire: :fire: Soon

:fire: :fire: Next

:fire: Maybe
- [ ] Server-based processing
- [ ] Multiprocessing options for speed-up processing

:checkered_flag: Done
- [x] Support for [Oriented Bounding Boxes (OBB) models from Ultralytics](https://docs.ultralytics.com/tasks/obb/?h=oriented) #184 
- [x] Support for [YOLOv9](https://github.com/WongKinYiu/yolov9) models #156  
- [x] Multihead support for regression and segmentation models #150
- [x] Speed-up global NMS for object detection results with cKDTree #147

**Comment by omranamini** [comment]
Hello,

I am encountering a MemoryError when trying to run the Deepness plugin in QGIS with a satellite image. The error message is:
Error! Processing error: "Unhandled exception occurred. See Python Console for details"  

Details: MemoryError - Unable to allocate 118. GiB for an array with shape (205432, 205432, 3) and data type uint8  
Traceback (most recent call last):  
  File "...deepness\processing\map_processor\map_processor.py", line 189, in _get_array_or_mmapped_array  
    full_result_img = np.zeros(final_shape_px, np.uint8)

The image I am using is a 4-band satellite image with 2-meter resolution. It seems that Deepness tries to allocate an extremely large array, which exceeds the available system memory.
Could you please advise on:
Whether there is a recommended workflow for processing very large images to avoid this memory issue.
Any built-in tiling or memory-efficient options in Deepness for handling large images.
Thank you for your support!
Best regards,
Omran Amini
<img width="1219" height="779" alt="123" src="https://github.com/user-attachments/assets/df7668ce-33cf-4422-be27-fdf2dac8bd68" />

**Comment by przemyslaw-aszkowski** [comment]
Hi @omranamini,

For the segmentation models, I think the entire output layer is created in RAM memory. At the end we need to have this layer in full size, because that how the segmentation works. With detection models it is simpler, as we do not have a specific value assigned to each pixel.

There is an idea we had in the past to use disk-backed numpy array (so the array will be stored on the SDD/HDD insted of RAM). It should be relatively easy to change in code. Let us know if you are really interested.

Cheers
Przemek
